### PR TITLE
VTTablet: smartconnpool: notify all expired waiters

### PR DESCRIPTION
**Pull request by brendar**
## Description

Fixes a bug that causes `waitlist.expire` to only remove the first expired waiter on each invocation. This is due to calling `wl.list.Remove(e)` while traversing the list. This sets `e.Next` to `nil`, which terminates the loop. The fix is to remove waiters from the list after traversing it. 

I think this should be backported to v19+. With the right combination of client behavior, query timeouts, and transaction timeouts, this bug can lead to tablets locking up indefinitely with all attempts to get a connection timing out.

## Related Issue(s)

Fixes https://github.com/vitessio/vitess/issues/16896

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

## Deployment Notes

n/a

### CODE REVIEW

**Review by mattlord — APPROVED** [review]
Thanks, @brendar ! ❤️

@vmg if you have a moment could you also look please?

File: go/pools/smartconnpool/waitlist_test.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+	// Expire the waiters
+	wait.expire(false)
+
+	// Wait for the notified goroutines to finish
+	timeout := time.After(1 * time.Second)
+	for expireCount.Load() != int32(waiterCount) {
+		select {
+		case <-timeout:
+			t.Fatalf("Timed out waiting for all waiters to expire. Wanted %d, got %d", waiterCount, expireCount.Load())
```
Comment (mattlord) [inline go/pools/smartconnpool/waitlist_test.go:58]: Nit, but We should use require.Fail* in new tests

File: go/pools/smartconnpool/waitlist_test.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+		case <-time.After(10 * time.Millisecond):
```
Comment (mattlord) [inline go/pools/smartconnpool/waitlist_test.go:59]: Nit, but slightly more efficient to use a ticker here

File: go/pools/smartconnpool/waitlist_test.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import (
+	"context"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestWaitlistExpireWithMultipleWaiters(t *testing.T) {
```
Comment (mattlord) [inline go/pools/smartconnpool/waitlist_test.go:28]: I'm assuming that this test fails on main. Just double checking 🙂

**Review by mattlord — APPROVED** [review]

File: go/pools/smartconnpool/waitlist_test.go (hunk 3 above)
Comment (brendar) [reply go/pools/smartconnpool/waitlist_test.go:29]: Yup, looks like this:
```
=== RUN   TestWaitlistExpireWithMultipleWaiters
    waitlist_test.go:61:
        	Error Trace:	/Users/brendan/src/github.com/Shopify/vitess/go/pools/smartconnpool/waitlist_test.go:61
        	Error:      	Timed out waiting for all waiters to expire
        	Test:       	TestWaitlistExpireWithMultipleWaiters
        	Messages:   	Wanted 2, got 1
```

**Review by vmg — APPROVED** [review]
Yeah this is correct. I think this is the cause of the reports we've seen of connections being stuck when shutting down the tablets and that we've never reproduced. Good catch!

**Review by harshit-gangal — COMMENTED** [review]
Good find!

**Pull request merged**